docs/docs/reference/metaprogramming/quote-matching.md

```diff
@@ -0,0 +1,64 @@
+---
+
+## Evaluation sematics
+
+Bellow we provide an informal description of the pattern evailatuion.
+We use `'{..}` for expression, `'[..]` for types and `{{..}}` for patterns nested in expressions.
+The semantics are defined as a list of reduction rules that are tried one by one until one matches.
+
+Operations:
+* `s =?= p` checks if a scrutinee `s` matches the pattern `p` while accumulating extracted parts of the code.
+* `isColosedUnder(x1, .., xn)('{e})` returns true if and only if all the references in `e` to names defined in the patttern are contained in the set `{x1, ... xn}`.
+* `lift(x1, .., xn)('{e})` returns `(y1, ..., yn) => [xi = $yi]'{e}` where `yi` is an `Expr` of the type of `xi`.
+* `withEnv(x1 -> y1, ..., xn -> yn)(matching)` evaluates mathing recording that `xi` is equivalent to `yi`.
+* `matched` denotes that the the match succedded and `matched('{e})` denotes that a matech succeded and extracts `'{e}`
+* `&&&` matches if both sides match. Concatenates the extracted expressions of both sides.
+
+```scala
+'{ e } =?= '{ $x }  &&  typeOf('{e}) <:< typeOf('{$x}) && isColosedUnder()('{e})  ===>   matched('{e})
+'{ e } =?= '{ $y(x1, ..., xn) }  &&  isColosedUnder(x1, ... xn)('{e})  ===>   matched(lift(x1, ..., xn)('{e}))
+
+'{ lit } =?= '{ lit }   ===>   matched
+'{ e: T } =?= '{ p: P }   ===>   '{e} =?= '{p} &&& '[T] =?= '[P]
+'{ e } =?= '{ p: P }   ===>   '{e} =?= '{p}
+'{ e.x } =?= '{ p.x }   ===>   '{e} =?= '{p}
+'{ x } =?= '{ x }   ===>   matched
+'{e0.f(e1, ..., en)} =?= '{p0.f(p1, ..., p2)}   ===>   '{e0} =?= '{p0} &&& '{e1} =?= '{p1} &&& ... %% '{en} =?= '{pn}
+'{e.f[T1, ..., Tn]} =?= '{p.f[P1, ..., Pn]}   ===>   '{e} =?= '{p} &&& '[T1] =?= '{P1} &&& ... %% '[Tn] =?= '[Pn]
+
+'{ if e0 then e1 else e2 } =?= '{ if p0 then p1 else p2 }   ===>  '{e0} =?= '{p0} &&& '{e1} =?= '{p1} &&& '{e2} =?= '{p2}
+'{ while e0 do e1 } =?= '{ while p0 do p1 }   ===>  '{e0} =?= '{p0} &&& '{e1} =?= '{p1}
+
+'{ e0 = e1 } =?= '{ p0 = p1 } && '{e0} =?= '{p0}   ===>   '{e1} =?= '{p1}
+'{ new T } =?= '{ new T }   ===>   matched
+'{ this } =?= '{ this }   ===>   matched
+'{ e.super[T] } =?= '{ p.super[T] }   ===>   '{e} =?= '{p}
+'{ Seq(e0, ..., en): _* } =?= '{ Seq(p0, ..., pn): _* }   ===>   '{e0} =?= '{p0} &&& ... &&& '{en} =?= '{pn}
+
+'[T] =?= '[P] && T <:< P   ===>   matched
+'[ T0[T1, ..., Tn] ] =?= '[ P0[P1, ..., Pn] ]   ===>   '[T0] =?= '[P0] &&& ... &&& '[Tn] =?= '[Pn]
+'[T @annot] =?= '[P]   ===>   '[T] =?= '[P]
+'[T] =?= '[P @annot]   ===>   '[T] =?= '[P]
+
+'{ {e0; e1; ...; en}; em } =?= '{ {p0; p1; ...; pm}; em }   ===>   '{ e0; {e1; ...; en; em} } =?= '{ p0; {p1; ...; pm; em} }
+'{ val x: T = e1; e2 } =?= '{ val y: P = p1; p2 }   ===>   withEnv(x -> y)('[T] =?= '[P] &&& '{e1} =?= '{p1} &&& '{e2} =?= '{p2})
+'{ def x0(x1: T1, ..., xn: Tn): T0 = e1; e2 } =?= '{ def y0(y1: P1, ..., yn: Pn): P0 = p1; p2 }   ===>   withEnv(x0 -> y0, ..., xn -> yn)('[T0] =?= '[P0] &&& ... &&& '[Tn] =?= '[Pn] &&& '{e1} =?= '{p1} &&& '{e2} =?= '{p2})
```

Comment:
The semantics for local definitions is surprising. I see the intention is to avoid scope extrusion, but it seems to be a big usability penalty here.

Reply:
@liufengyun What alternative semantics are you thinking about?

Reply:
The original comment is actually about the following rule:

> '{ e } =?= '{ $x }  &&  typeOf('{e}) <:< typeOf('{$x}) && isColosedUnder()('{e})  ===>  
 matched('{e})

From my limited meta-programming experience, I like quoted pattern matching on terms and types a lot, it's a big productivity booster. In particular, it saves a lot of boilerplate when pattern matching on known methods and types. This makes developing macro-based DSLs a joyful experience. Previously, they have to be performed via low-level reflect API in a tedious way or doing it in an unsafe way via string comparison.  @nicolasstucki has done a great job here for Scala meta-programmers👍 

However, I'm not sure that making it _safer_ and feature-complete will help much in practice, in particular when it comes at the price of usability. By usability, I mean simple API, natural and precise semantics, stability, debuggability, a clear-cut boundary of what is supported (accountable & trustable) and what is not.

For example, maybe it is safe drop the support for quoted code patterns of pattern matches and definitions, and leave them to reflect API?